src/chainrules.jl

```diff
@@ -27,7 +27,7 @@ function ChainRulesCore.rrule(
         ::typeof(*),
         bm::BlockDiagonal{T, V},
         v::StridedVector{T}
-    ) where {T<:Union{Real, Complex}, V<:Matrix{T}}
+    ) where {T<:Union{Real, Complex}, V}
```

Comment:
We should continue to maintain strict type constraints here as this rule is not in general correct for blocks that aren't `Matrix{<:Real}`s (or nearly-`Matrix{<:Real}`s, like `StridedMatrix{<:Real}`s). Something like
```julia
V<:Vector{<:Matrix{<:Real}}
```
would actually be a good choice to my mind. See e.g. https://github.com/JuliaDiff/ChainRules.jl/issues/232

Reply:
love it. makes this PR much easier 😂